src/main/python/ayab/preferences.py

```diff
@@ -35,15 +36,37 @@
 from .machine import Machine
 from .language import Language
 from .scene import AspectRatio
+from typing import TYPE_CHECKING, Any, Callable, Literal, Optional, TypeAlias, TypeVar, TypedDict, cast 
+if TYPE_CHECKING:
+    from .ayab import GuiMain
 
+T = TypeVar('T')
 
-def str2bool(qvariant):
+def str2bool(qvariant:str|bool)->bool:
     if type(qvariant) == str:
         return qvariant.lower() == "true"
     else:
-        return qvariant
+        return cast(bool,qvariant)
 
 
+
+PreferencesDictBoolKeys: TypeAlias = Literal['default_infinite_repeat', 'default_knit_side_image', 'quiet_mode', 'disable_hardware_beep']
+PreferencesDictObjKeys: TypeAlias = Literal['aspect_ratio', 'default_alignment', 'default_knitting_mode', 'machine']
+PreferencesDictKeys : TypeAlias = Literal[PreferencesDictBoolKeys,PreferencesDictObjKeys,'language']
+
+PreferencesDict = TypedDict('PreferencesDict',{
+    'machine':type[Machine],
+    'default_knitting_mode': type[Mode],
+    'default_infinite_repeat': type[bool],
+    'default_alignment': type[Alignment],
+    'default_knit_side_image': type[bool],
+    'aspect_ratio': type[AspectRatio],
+    #'default_continuous_reporting': bool,
+    'quiet_mode': type[bool],
+    'disable_hardware_beep': type[bool],
+    'language': type[Language],
+})
```

Comment:
This is the most effort I put into typing anything- I think it's important to make sure all of this is clearly defined so that later functions can leverage it effectively.